File: frontend/src/utils/croFreeTime.ts
import { todayIsoDate } from './datetime'

const MONTHS: Record<string, number> = {
  jan: 1,
  feb: 2,
  mar: 3,
  apr: 4,
  may: 5,
  jun: 6,
  jul: 7,
  aug: 8,
  sep: 9,
  oct: 10,
  nov: 11,
  dec: 12,
}

/** Normalize CRO / API free-time strings (yyyy-MM-dd or dd-MMM-yyyy) to ISO date. */
export function parseCroFreeTimeToIso(value: string | null | undefined): string | null {
  if (!value?.trim()) return null
  const raw = value.trim()

  if (/^\d{4}-\d{2}-\d{2}$/.test(raw)) return raw

  const mmm = /^(\d{1,2})-([A-Za-z]{3})-(\d{4})$/.exec(raw)
  if (mmm) {
    const day = Number(mmm[1])
    const month = MONTHS[mmm[2].toLowerCase()]
    const year = Number(mmm[3])
    if (!month || day < 1 || day > 31) return null
    return `${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`
  }

  return null
}

/** True when CRO free demurrage date is before today (Philippines calendar). */
export function isCroFreeTimeExpired(value: string | null | undefined): boolean {
  const iso = parseCroFreeTimeToIso(value)
  return Boolean(iso && iso < todayIsoDate())
}

export function croFreeTimeExpiredMessage(freeUntil: string | null | undefined): string {
  const iso = parseCroFreeTimeToIso(freeUntil)
  const when = iso ?? freeUntil ?? 'the free-time date'
  return `Free demurrage time expired on ${when}. You can still save a draft, but demurrage and detention charges must be settled before you submit this pre-forecast.`
}

File: android-trucker/app/src/main/java/com/ecms/trucker/util/CroFreeTime.kt
package com.ecms.trucker.util

import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

private val monthMap = mapOf(
    "jan" to 1, "feb" to 2, "mar" to 3, "apr" to 4, "may" to 5, "jun" to 6,
    "jul" to 7, "aug" to 8, "sep" to 9, "oct" to 10, "nov" to 11, "dec" to 12,
)

fun parseCroFreeTimeToIso(value: String?): String? {
    val raw = value?.trim().orEmpty()
    if (raw.isEmpty()) return null

    if (Regex("""^\d{4}-\d{2}-\d{2}$""").matches(raw)) return raw

    val mmm = Regex("""^(\d{1,2})-([A-Za-z]{3})-(\d{4})$""").find(raw)
    if (mmm != null) {
        val day = mmm.groupValues[1].toIntOrNull() ?: return null
        val month = monthMap[mmm.groupValues[2].lowercase(Locale.US)] ?: return null
        val year = mmm.groupValues[3].toIntOrNull() ?: return null
        return LocalDate.of(year, month, day).format(DateTimeFormatter.ISO_LOCAL_DATE)
    }
    return null
}

fun isCroFreeTimeExpired(value: String?): Boolean {
    val iso = parseCroFreeTimeToIso(value) ?: return false
    return iso < LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE)
}

fun croFreeTimeExpiredMessage(value: String?): String {
    val until = value?.trim().orEmpty().ifEmpty { "the stated date" }
    return "CRO/eDO free time expired on $until. Settle charges under Demurrage before submitting."
}
